### ISSUE TITLE
Add -report to Scan

### ISSUE THREAD

**Issue by brurend** [fastlane/fastlane#4327]
### New Issue Checklist

- [X] Updated fastlane to the latest version
- [X] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

Not really an issue, its more of a suggestion. Would be great if Scan could generate  a json-compilation-database report, in a similar way to xctool to use with oclint/swiftlint.

Also fastlane action xctool is not showing any available options, not sure if this is intended or a bug.

### Environment

fastlane version (run `fastlane -v`): 1.81.0

Do you use bundler to execute fastlane (i.e. `bundle exec fastlane`)?
No
Do you use a Ruby environment manager (e.g. `chruby`, `rbenv`, `rvm`)?
rbenv

**Comment by TKBurner** [comment]
@brurend Great suggestion. I'm going to mark this as help wanted in case anyone from the community wants to work on it. 👍

**Comment by andybest** [comment]
It seems that scan can currently generate the report by using the following

```
scan(output_types: "json-compilation-database")
```

The problem I'm having is that the OCLint action requires the compile command file to be specified so it can get the list of files from it in order to build the OCLint commands. It then passes the parameter directly to OCLint, which requires the directory containing the compilation database as the parameter, not a file.

**Comment by TKBurner** [comment]
Thanks for sharing your config @andybest! Are you interested in submitting a PR to create the desired behavior? Thanks 👍

**Comment by andybest** [comment]
@TKBurner sure, I'm not sure what the best solution to the naming problem is though. At the moment, I'm renaming the compilation database file after Scan generates it, then using a custom method based on the OCLint action to run it:

```
fastlane_version "1.89.0"

default_platform :ios

platform :ios do
  desc "Runs all the tests"
  lane :test do
    scan(
        clean: true,
        scheme: "MyVW",
        output_types: "junit,html,json-compilation-database"
    )
    
    run_oclint(Dir.pwd + "/test_output/report.json-compilation-database")
  end

def run_oclint(compile_commands_file)
  # The OCLint action in fastlane is not currently working properly.
  # This is a workaround until it is fixed.
  
  # Parameters
  report_output = File.dirname(compile_commands_file) + "/oclint.pmd"
  
  select_regex = /.*/
  exclude_regex = /Pods/
  
  thresholds = [
      'LONG_LINE=200',
      'LONG_METHOD=200'
  ]
  
  enable_rules = [
    
  ]
  
  disable_rules = [
    
  ]
  
  enable_static_analyzer = false
  enable_global_analysis = false
  allow_duplicated_violations = false
  
  # Copy the file to the correct location
  FileUtils.cp(compile_commands_file, File.dirname(compile_commands_file) + "/compile_commands.json")
  
  # Get file list
  files = JSON.parse(File.read(compile_commands_file)).map do |compile_command|
    file = compile_command['file']
    File.exist?(file) ? file : File.join(compile_command['directory'], file)
  end
  
  files.uniq!
  files.select! do |file|
    file_ruby = file.gsub('\ ', ' ')
    File.exist?(file_ruby) and
      (!select_regex or file_ruby =~ select_regex) and
      (!exclude_regex or !(file_ruby =~ exclude_regex))
  end
  
  oclint_args = ["-report-type=pmd", "-o=#{report_output}"]
  
  oclint_args << thresholds.map { |t| "-rc=#{t}" }
  
  # Escape ' in rule names with \' when passing on to shell command
  oclint_args << enable_rules.map { |r| "-rule #{r.shellescape}" } if enable_rules.length > 0
  oclint_args << disable_rules.map { |r| "-disable-rule #{r.shellescape}" } if disable_rules.length > 0

  oclint_args << "-enable-clang-static-analyzer" if enable_static_analyzer
  oclint_args << "-enable-global-analysis" if enable_global_analysis
  oclint_args << "-allow-duplicated-violations" if allow_duplicated_violations
  oclint_args << "-p #{File.dirname(compile_commands_file)}"
  
  command = [
    "oclint",
    oclint_args,
    '"' + files.join('" "') + '"'
  ].join(' ')
  
  sh(command)
  
end
```

Ideally, you'd be able to specify the output file name from scan, so that the filename doesn't have to be manually changed.

**Comment by andybest** [comment]
@TKBurner sure, I'm not sure what the best solution to the naming problem is though. At the moment, I'm renaming the compilation database file after Scan generates it, then using a custom method based on the OCLint action to run it:

```
fastlane_version "1.89.0"

default_platform :ios

platform :ios do
  desc "Runs all the tests"
  lane :test do
    scan(
        clean: true,
        scheme: "MyApp",
        output_types: "junit,html,json-compilation-database"
    )
    
    run_oclint(Dir.pwd + "/test_output/report.json-compilation-database")
  end

def run_oclint(compile_commands_file)
  # The OCLint action in fastlane is not currently working properly.
  # This is a workaround until it is fixed.
  
  # Parameters
  report_output = File.dirname(compile_commands_file) + "/oclint.pmd"
  
  select_regex = /.*/
  exclude_regex = /Pods/
  
  thresholds = [
      'LONG_LINE=200',
      'LONG_METHOD=200'
  ]
  
  enable_rules = [
    
  ]
  
  disable_rules = [
    
  ]
  
  enable_static_analyzer = false
  enable_global_analysis = false
  allow_duplicated_violations = false
  
  # Copy the file to the correct location
  FileUtils.cp(compile_commands_file, File.dirname(compile_commands_file) + "/compile_commands.json")
  
  # Get file list
  files = JSON.parse(File.read(compile_commands_file)).map do |compile_command|
    file = compile_command['file']
    File.exist?(file) ? file : File.join(compile_command['directory'], file)
  end
  
  files.uniq!
  files.select! do |file|
    file_ruby = file.gsub('\ ', ' ')
    File.exist?(file_ruby) and
      (!select_regex or file_ruby =~ select_regex) and
      (!exclude_regex or !(file_ruby =~ exclude_regex))
  end
  
  oclint_args = ["-report-type=pmd", "-o=#{report_output}"]
  
  oclint_args << thresholds.map { |t| "-rc=#{t}" }
  
  # Escape ' in rule names with \' when passing on to shell command
  oclint_args << enable_rules.map { |r| "-rule #{r.shellescape}" } if enable_rules.length > 0
  oclint_args << disable_rules.map { |r| "-disable-rule #{r.shellescape}" } if disable_rules.length > 0

  oclint_args << "-enable-clang-static-analyzer" if enable_static_analyzer
  oclint_args << "-enable-global-analysis" if enable_global_analysis
  oclint_args << "-allow-duplicated-violations" if allow_duplicated_violations
  oclint_args << "-p #{File.dirname(compile_commands_file)}"
  
  command = [
    "oclint",
    oclint_args,
    '"' + files.join('" "') + '"'
  ].join(' ')
  
  sh(command)
  
end
```

Ideally, you'd be able to specify the output file name from scan, so that the filename doesn't have to be manually changed.

**Comment by andybest** [comment]
I have provided a method to integrate Scan and OCLint with this PR #4841

**Comment by TKBurner** [comment]
@andybest Thanks for your PR and for providing test coverage. The team is going to review it and work with you. Thanks! 👍

**Comment by mfurtak** [comment]
We've addressed this with https://github.com/fastlane/fastlane/pull/4992, which will go out with the next release 🚀

**Comment by mfurtak** [comment]
This is released as part of [`scan` 0.8.0](https://github.com/fastlane/fastlane/releases/tag/scan%2F0.8.0) 🎉
